pkg/provider/provider.go

```diff
@@ -93,6 +85,115 @@ func configure(version string, p *schema.Provider) func(context.Context, *schema
 		// api.BuildAPI(meta).RegionClient()
 
 		userAgent := fmt.Sprintf("%s/%s", "terraform-provider-biganimal", version)
+		diags := diag.Diagnostics{}
+
+		if ba_bearer_token == "" {
+			ba_bearer_token = os.Getenv("BA_BEARER_TOKEN")
+		}
+
+		if ba_bearer_token == "" {
+			diags = append(diags, diagv2.Diagnostic{
+				Severity: diagv2.Error,
+				Summary:  "Unable to find ba_bearer_token",
+				Detail:   "ba_bearer_token cannot be an empty string"})
+			return nil, diags
+		}
+
+		if ba_api_uri == "" {
+			ba_api_uri = os.Getenv("BA_API_URI")
+		}
+		if ba_api_uri == "" {
+			ba_api_uri = "https://portal.biganimal.com/api/v3"
+		}
+
 		return api.NewAPI(ba_bearer_token, ba_api_uri, userAgent), nil
 	}
 }
+
+type bigAnimalProvider struct {
+	version string
+}
+
+// providerData can be used to store data from the Terraform configuration.
+type providerData struct {
+	BaBearerToken *string `tfsdk:"ba_bearer_token"`
+	BaAPIUri      *string `tfsdk:"ba_api_uri"`
+}
+
+func NewProvider(version string) func() provider.Provider {
+	return func() provider.Provider {
+		return &bigAnimalProvider{
+			version: version,
+		}
+	}
+}
+
+func (b bigAnimalProvider) Metadata(_ context.Context, _ provider.MetadataRequest, resp *provider.MetadataResponse) {
+	resp.TypeName = "biganimal"
+	resp.Version = b.version
+}
+
+func (b bigAnimalProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
```

Comment:
Can you write unit tests for this Configure function, to ensure that we cover all the cases. It would be helpful to check cases like:
1. Those two fields are set in the provider block
2. One of them are set in the provider block
3. They are not set in the provider block, but present as environment variables
4. no env vars, not defined within a provider block, we expect diag error.